File: docs/CUForge/terra-current-use-backend-phase1/terra-current-use-backend-phase1/backend/TerraFusion.Modules.CurrentUse/Domain/Rollback/RollbackRules.cs
using TerraFusion.Modules.CurrentUse.Domain;

namespace TerraFusion.Modules.CurrentUse.Domain.Rollback;

public static class RollbackRules
{
    public const string CalculationVersion = "CU_ROLLBACK_ENGINE_v2026_03_01";

    private static readonly DateOnly FarmAgFourYearEffectiveDate = new(2025, 9, 1);

    public static int DetermineRollbackYearCount(ClassificationType classificationType, DateOnly removalDate)
    {
        if (classificationType == ClassificationType.FarmAndAgricultural &&
            removalDate >= FarmAgFourYearEffectiveDate)
        {
            return 4;
        }

        return classificationType switch
        {
            ClassificationType.OpenSpace => 7,
            ClassificationType.TimberLand => 7,
            ClassificationType.FarmAndAgricultural => 7,
            ClassificationType.DesignatedForestland => 7,
            _ => 7
        };
    }

    public static IReadOnlyList<int> BuildRollbackYears(int taxYearOfRemoval, int rollbackYearCount)
    {
        return Enumerable
            .Range(1, rollbackYearCount)
            .Select(offset => taxYearOfRemoval - offset)
            .OrderBy(year => year)
            .ToArray();
    }

    public static bool ShouldSuppressPenalty(
        PenaltySuppressionReason? penaltySuppressionReason,
        StatutoryExceptionReason? statutoryExceptionReason)
    {
        if (statutoryExceptionReason is not null and not StatutoryExceptionReason.None)
        {
            return true;
        }

        return penaltySuppressionReason is
            PenaltySuppressionReason.QualifyingVoluntaryWithdrawal or
            PenaltySuppressionReason.StatutoryException;
    }
}

File: docs/CUForge/terra-current-use-phase1/terra-current-use-phase1/src/modules/terra-current-use/domain/rollback/rollbackRules.ts
import type { ClassificationType } from '../../types/currentUseTypes';
import type { RollbackCalculationInput } from './rollbackTypes';

export const CURRENT_USE_ROLLBACK_ENGINE_VERSION = 'CU_ROLLBACK_ENGINE_v2026_03_01';

const FARM_AG_FOUR_YEAR_EFFECTIVE_DATE = new Date('2025-09-01T00:00:00.000Z');

export function determineRollbackYearCount(
  classificationType: ClassificationType,
  removalDateIso: string,
): number {
  const removalDate = new Date(removalDateIso);

  if (
    classificationType === 'FARM_AND_AGRICULTURAL' &&
    removalDate >= FARM_AG_FOUR_YEAR_EFFECTIVE_DATE
  ) {
    return 4;
  }

  if (classificationType === 'OPEN_SPACE' || classificationType === 'TIMBER_LAND') {
    return 7;
  }

  if (classificationType === 'FARM_AND_AGRICULTURAL') {
    return 7;
  }

  if (classificationType === 'DESIGNATED_FORESTLAND') {
    return 7;
  }

  return 7;
}

export function buildRollbackYears(taxYearOfRemoval: number, rollbackYearCount: number): number[] {
  return Array.from({ length: rollbackYearCount }, (_, index) => taxYearOfRemoval - 1 - index).sort();
}

export function shouldSuppressPenalty(input: RollbackCalculationInput): boolean {
  if (input.statutoryExceptionReason && input.statutoryExceptionReason !== 'NONE') return true;
  if (input.penaltySuppressionReason === 'QUALIFYING_VOLUNTARY_WITHDRAWAL') return true;
  if (input.penaltySuppressionReason === 'STATUTORY_EXCEPTION') return true;
  return false;
}
